### PR TITLE
Inheritance based implementation for user code

### LINKED ISSUE #54: Interface for developers to implement should be an actual interface
The current API for game devs to access is a variable in a namespace, however this could be changed to be an actual interface (i.e. pure virtual) plus class factory to register the implemented class.

### PR DESCRIPTION
**Pull request by Nanogamer7**
See issue #54

### CODE REVIEW

**Comment by Nanogamer7** [comment]
New file names might not be ideal, especially as they reuse old names for different purposes. This creates issues with git diffs

File: include/GameFactory.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    CreateFunc creator;
+    Game* gameInstance = nullptr; //!< \depricated
+    // std::map<std::string, CreateFunc> creators;
+};
+
+#define REGISTER_GAME(className) \
+    static bool registered_##className = []() { \
+        GameFactory::instance().registerClass([]() -> Game* { return new className; }); \
+        return true; \
+    }();
```
Comment (Nimac0) [inline include/GameFactory.h]: macro bad, go into the shame corner >:c

**Review by Nimac0 — CHANGES_REQUESTED** [review]

File: include/GameFactory.h (hunk 1 above)
Comment (Nanogamer7) [reply include/GameFactory.h]: done (the shame part)

please review again
also check https://github.com/VEGO-Engine/chickengame/tree/beni

Comment (Nimac0) [reply include/GameFactory.h]: lgtm

**Review by Nimac0 — APPROVED** [review]

**Pull request merged**